### ISSUE TITLE
Create `@xw.cached` decorator to generate VBA code to cache UDF results

### ISSUE THREAD

**Issue by stenci** [xlwings/xlwings#1981]
### Caching return values could make the UDFs much faster
In workbooks with thousands of calls to xlwings' UDFs where most calls use the same argument values, 99% of the calculation time is wasted running the same functions with the same values again and again.

Creating a decorator `@xw.cached` that does nothing in Python and it's used only to add a few lines to the VBA code generated when clicking on `Import Functions`, would prevent the re-running the same functions by getting the values from a dictionary.

### How I add caching to my xlwings UDFs
I don't understand how the `Import Functions` button works. I think it calls a dll written in C++ and I don't speak C++, so I can't help on that front.

I created a Jupyter notebook that helps me with adding the caching code to the `xlwings_udfs` code and it seems to be working fine. Now I can use xlwings UDFs on worksheets that call thousands of times UDFs defined in Python, as long as the unique set of arguments values is very small.

My workflow is:
* Define the functions on the `udfs.py` file
* Click on the `Import Functions` button
* Select all the code from the `xlwings_udfs` module
* Paste it on the Jupyter cell that defines the `vba_code` variable
* Create one cell with one call to `add_caching()` for each UDF I want to cache
* Copy the resulting `vba_code`
* Paste it back into `xlwings_udfs` and replace the old code

At this point when an UDF with cache is called, it will first check if a call with the same set of parameters has already been done. If yes, then it will return the cached value, if no, then it will call the Python function, store the value in cache and return it.

### The Jupyter notebook
Here is an example of the 3 cells of one Jupyter notebook that I have just used:
```Python
# this cell assigns a string with the content of the xlwings_udfs module to the vba_code variable 
vba_code = """
'Autogenerated code by xlwings - changes will be lost with next import!
#Const App = "Microsoft Excel" 'Adjust when using outside of Excel
Function u110_equitone_horizontal_span_2_holes(wind_load, framing_type, framing_spacing)
    #If App = "Microsoft Excel" Then
        If TypeOf Application.Caller Is Range Then On Error GoTo failed
[...]
        Exit Function
    #End If
failed:
    u110_equitone_vertical_span_3_holes = Err.Description
End Function
"""
```

```Python
# this cell defines the add_caching() function used to add the caching code to vba_code
import re

def add_caching(vba_code, function_name):
    dictionary_exists = False
    dictionary = 'Dim UdfCache As Dictionary'
    definition = f'Function {function_name}('
    call_udf = f'        {function_name} = Py.CallUDF("'
    
    lines = vba_code.split('\n')
    for n_line, line in enumerate(lines):
        dictionary_exists = dictionary_exists or line == dictionary

        if line.startswith(definition):
            if lines[n_line + 1] == f'    Dim Key As String':
                continue

            arguments = re.findall(r'(?:(?:, )|\()(\w+)', line)
        
            lines.insert(n_line + 1, f'    Dim Key As String')
            lines.insert(n_line + 2, f'    Key = "{function_name}"')
            
            n_arg = 0
            for n_arg, argument in enumerate(arguments):
                lines.insert(n_line + n_arg + 3, f'    Key = Key & VbTab & {argument}')
                
            lines.insert(n_line + n_arg + 4, f'    If UdfCache Is Nothing Then Set UdfCache = NewDictionary')
            lines.insert(n_line + n_arg + 5, f'    If UdfCache.Exists(Key) Then')
            lines.insert(n_line + n_arg + 6, f'        {function_name} = UdfCache(Key)')
            lines.insert(n_line + n_arg + 7, f'        Exit Function')
            lines.insert(n_line + n_arg + 8, f'    End If')

        if line.startswith(call_udf):
            if lines[n_line + 1] == f'        UdfCache.Add Key, {function_name}':
                continue

            lines.insert(n_line + 1, f'        UdfCache.Add Key, {function_name}')

    if not dictionary_exists:
        lines.insert(3, '')
        lines.insert(4, dictionary)
        lines.insert(5, '')

    return '\n'.join(lines)
```

```Python
# this cell uses add_caching() to add the caching code to all the functions that require caching
vba_code = add_caching(vba_code, 'u110_equitone_max_panel_width_1_hole')
vba_code = add_caching(vba_code, 'u110_equitone_horizontal_span_2_holes')
vba_code = add_caching(vba_code, 'u110_equitone_horizontal_span_3_holes')
vba_code = add_caching(vba_code, 'u110_equitone_max_panel_height_1_hole')
vba_code = add_caching(vba_code, 'u110_equitone_vertical_span_2_holes')
vba_code = add_caching(vba_code, 'u110_equitone_vertical_span_3_holes')
print(vba_code)
```

This is an example of the resulting code:
```VBA
'Autogenerated code by xlwings - changes will be lost with next import!
#Const App = "Microsoft Excel" 'Adjust when using outside of Excel

Dim UdfCache As Dictionary

Function u110_equitone_horizontal_span_2_holes(wind_load, framing_type, framing_spacing)
    Dim Key As String
    Key = "u110_equitone_horizontal_span_2_holes"
    Key = Key & vbTab & wind_load
    Key = Key & vbTab & framing_type
    Key = Key & vbTab & framing_spacing
    If UdfCache Is Nothing Then Set UdfCache = NewDictionary
    If UdfCache.Exists(Key) Then
        u110_equitone_horizontal_span_2_holes = UdfCache(Key)
        Exit Function
    End If
    #If App = "Microsoft Excel" Then
        If TypeOf Application.Caller Is Range Then On Error GoTo failed
        u110_equitone_horizontal_span_2_holes = Py.CallUDF("stefano", "u110_equitone_horizontal_span_2_holes", Array(wind_load, framing_type, framing_spacing), ThisWorkbook, Application.Caller)
        UdfCache.Add Key, u110_equitone_horizontal_span_2_holes
        Exit Function
    #Else
        u110_equitone_horizontal_span_2_holes = Py.CallUDF("stefano", "u110_equitone_horizontal_span_2_holes", Array(wind_load, framing_type, framing_spacing))
        UdfCache.Add Key, u110_equitone_horizontal_span_2_holes
        Exit Function
    #End If
failed:
[...]
```

Since `Dictionary` is not creatable, I have added the following function to xlwings.utils:
```VBA
Function NewDictionary(ParamArray Args()) As Dictionary
  Dim I As Integer
  Set NewDictionary = New Dictionary
  For I = 0 To UBound(Args) Step 2
    NewDictionary.Add Args(I), Args(I + 1)
  Next I
End Function
```

This is an old function of mine, it can be used in the form `NewDictionary("Key1", value1, "Key2", value2, ...)`, `xlwings` doesn't use this form, but I left it there, just in case it will turn useful in the future.

### Note about the dictionary key
The key for the cache dictionary is generated by converting the function arguments to string. This works perfectly when the arguments are strings or integers, can do some very small rounding with floats, will not work with ranges.

On similar cached functions on other projects I convert the arguments to JSON (using https://github.com/VBA-tools/VBA-JSON) and use that as the dictionary key.

### A little background about my current use case
We produce panels to clad buildings. A building can have 1000+ panels. A sheet has the list of panels, with their size and other parameters and calculates the number of components, fasteners, etc. and calculates the cost for the entire project.

The number of fasteners for each panel depends on the type of fasteners, on the wind load and on the substrate.

The values of the parameters are very few on each spreadsheet. For example a building will have 2-3 wind load values depending on the distance of the panel from the floor and from the corners of the building. There will be 2-3 type of fasteners depending on aesthetical requirements that can change on different areas of the building.

For example we could have 300 panels with screws and 700 with rivets. 400 panels with 40 mph wind, 400 with 80 mph and 200 with 120 mph. The total is 3x2=6 different sets of arguments for 1000 panels. By caching the returned values I can reduce the number of calls from 1000 to 6.

Calculating the number of fasteners it in Excel would require an unmaintainable `LET()` function that defines 50+ temporary variables and uses tons of `IFS()`.

Calculating it in Python is a breeze, hence xlwings.

**Comment by ericbmoreira** [comment]
Does your Excel workbook stay open while you are doing all of these calculations? I'm asking because I run a financial application that uses Excel and xlwings, and it stays open for the entire trading day. I don't use Jupyter, I just coded it in VS Code and then made sure that I enabled "UDF Server". The UDF server will hold the python process open for the life of the Excel instance, and with that it will keep all global variables, etc.. I have a global dictionary that I access in a similar fashion to what you are talking about. I initialize my python code when I open the Excel workbook and then just run macros that call a python function, my VBA code is really basic because all it does is call the python function. I actually reduced the UDFs, from how many I had in the first place, and decided to just use Excel as a GUI by pasting the values into the Excel ranges.

Just an idea.

**Comment by stenci** [comment]
> The UDF server will hold the python process open for the life of the Excel instance, and with that it will keep all global variables, etc..

My Python functions are very fast, I'm not worried about Python taking too long doing its job. I'm worried about the time spent calling Python thousands of times.

> I have a global dictionary that I access in a similar fashion to what you are talking about

If I understand your global dictionary is in Python. My cache dictionary is in VBA, so I don't even need to get out of VBA when the functions have been called with the same values.

---

The user (estimator in this case) reads the building drawing and adds one line to the spreadsheet for each panel on the building drawing. They usually start with all panels with the same settings, then they change some settings for some groups of panels. For example they add 200 panels all with the same fasteners, then they change the fastener to 1 panel and copy it down to another 50 panels and end up with 50 panels with one fastener and 150 panels with another one.

Then they may decide to change the wind load for all panels in one shot, which would mean executing all the UDFs at once.

I am evaluating xlwings right now and, if it works, I will have 10-20 xlwings UDFs per panel. This means that changing the material (defined in one cell for all the panels) on a sheet with 5000 panels will trigger the execution of 100000 UDFs. Adding VBA side caching will reduce the 100000 calls down to a few dozens. This is the difference between being usable and not being usable.

---

I have tested UDF speed with `RunPython: Use UDF Server` both checked and unchecked, and I found out that when it's unchecked the UDFs run slightly faster.

This little test:
```VBA
Sub Test1()
  Dim T0 As Single
  T0 = Timer
  Application.CalculateFull
  Debug.Print Timer - T0
End Sub
```
executed with 170 UDF calls:
```text
''''' checked
 2.550781 
 2.566406 
 2.535156 

''''' unchecked
 2.441406 
 2.480469 
 2.4375 
```

**Comment by ericbmoreira** [comment]
Ok. I guess the way I am looking at it is that I would pass python all of the parameters, and do all of the calculations in the python app, rather than using s bunch of UDFs for each row. If there could just be user input that could change the various parameters, then it could be recalculated within python. It just sounds like a lot of UDFs. It sounds like it could be a pandas dataframe, with 1000's of panels (rows from Excel), which certain columns (wind load) could then be changed. Once it is recalculated, the values are pasted back into Excel.

**Comment by stenci** [comment]
I would love the full Python approach, but:

1. I'm working on templates with 40000 lines of VBA code, and I can't pull a switch and rewrite all overnight
2. User work on a few panels at a time and want to see the calculated values in real time

It's ok to wait a few seconds, maybe a minute, when a global value is changed and all the panels need to be recalculated, but the sheet must be responsive when working on one single panel.

I am thinking about replacing the hundreds of formulas associated with each panel with one single call to an xlwings function that updates all the panel values in one shot. The call would be triggered by some event, not by an UDF evaluation...

This would keep the application responsive when working on one panel and would allow a global refresh on demand by scanning all the panel rows.

Mmmmh...

Regardless, caching the values of xlwings functions would be a great improvement. 

For the time being I will keep adding the caching manually, hopefully one day I will see the `@xw.cached` decorator :)

**Comment by fzumstein** [comment]
I agree, it's a great idea, but I'd call it `xw.cache` to be in line with `functools.cache`. Also, you don't need C++ to change the VBA wrapper code, that's done here: https://github.com/xlwings/xlwings/blob/main/xlwings/udfs.py#L573

**Comment by antoinedeb** [comment]
Hello @fzumstein 

I wanted to know if you were still planning to implement the @xw.cache decorator in Xlwings.  I think it would be very useful. I saw that @stenci did a merge request with an implementation. have you planned to merge it into the package ? 
Thanks in advance for your help

**Comment by fzumstein** [comment]
Yes, this is still on the list, but I'd want to do it first in xlwings Server, as the calls there are even more expensive. Can't promise any ETA though.
